File: module-04-waiting/js/exponential_backoff.js
const { chromium } = require('playwright');

(async () => {
  const browser = await chromium.launch();
  const page    = await browser.newPage();
  await page.goto('https://example.com/ajax-data');

  // 1. 檢查條件函式
  async function dataLoaded() {
    const count = await page.$$eval('.data-item', els => els.length);
    return count >= 1;
  }

  // 2. 設定退避參數
  let delay       = 500;   // 毫秒
  const maxTrials = 4;

  // 3. 指數退避輪詢
  for (let i = 1; i <= maxTrials; i++) {
    if (await dataLoaded()) {
      console.log(`[JS] Data loaded after ${i} attempt(s)`);
      break;
    }
    console.log(`[JS] Attempt ${i} failed; sleeping ${delay}ms`);
    await page.waitForTimeout(delay);
    delay *= 2;  // 下一次等待時間 ×2
    if (i === maxTrials) {
      throw new Error("Data did not load within backoff attempts");
    }
  }

  // 4. 取得並輸出資料
  const items = await page.$$eval('.data-item', els => els.map(e => e.textContent));
  items.forEach((text, idx) => console.log(`Item ${idx+1}: ${text}`));

  await browser.close();
})();

File: module-04-waiting/py/exponential_backoff.py
import time
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from webdriver_manager.chrome import ChromeDriverManager
from selenium.webdriver.common.by import By

# 1. 啟動 headless Chrome
driver = webdriver.Chrome(
    service=Service(ChromeDriverManager().install()),
    options=webdriver.ChromeOptions().add_argument("--headless")
)
driver.get("https://example.com/ajax-data")

# 2. 定義檢查條件函式：判斷所需資料是否已加載
def data_loaded():
    # 假設 AJAX 會生成 .data-item 元素
    items = driver.find_elements(By.CSS_SELECTOR, ".data-item")
    return len(items) >= 1

# 3. 設定退避參數
initial_delay = 0.5     # 秒
max_attempts = 4        # 最多重試次數
delay = initial_delay

# 4. 指數退避輪詢
for attempt in range(1, max_attempts + 1):
    if data_loaded():
        print(f"[Python] Data loaded after {attempt} attempt(s)")
        break
    print(f"[Python] Attempt {attempt} failed; sleeping {delay}s before retry")
    time.sleep(delay)
    delay *= 2            # 延長等待時間（2 的指數）
else:
    raise TimeoutError("Data did not load within backoff attempts")

# 5. 成功後擷取資料
elements = driver.find_elements(By.CSS_SELECTOR, ".data-item")
for idx, el in enumerate(elements, 1):
    print(f"Item {idx}: {el.text}")

driver.quit()
